subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -385,256 +357,1118 @@ static int hfp_ag_send_data(struct bt_hfp_ag *ag, bt_hfp_ag_tx_cb_t cb, void *us
+{
+	bool is_inband_ringtone;
+
+	hfp_ag_lock(ag);
+	is_inband_ringtone = (ag->ag_features & BT_HFP_AG_FEATURE_INBAND_RINGTONE) ? true : false;
+	hfp_ag_unlock(ag);
+
+	if (is_inband_ringtone && !atomic_test_bit(ag->flags, BT_HFP_AG_INBAND_RING)) {
+		int err = hfp_ag_send_data(ag, NULL, NULL, "\r\n+BSIR:1\r\n");
+
+		atomic_set_bit_to(ag->flags, BT_HFP_AG_INBAND_RING, err == 0);
+	}
+}
+
+static void bt_hfp_ag_slc_connected(struct bt_hfp_ag *ag, void *user_data)
+{
+	bt_hfp_ag_set_state(ag, BT_HFP_CONNECTED);
+	(void)hfp_ag_next_step(ag, bt_hfp_ag_set_in_band_ring, NULL);
+}
+
+static int bt_hfp_ag_cmer_handler(struct bt_hfp_ag *ag, struct net_buf *buf)
+{
+	uint32_t number;
+	int err;
+	static const uint32_t command_line_prefix[] = {3, 0, 0};
+
+	if (!is_char(buf, '=')) {
+		return -ENOTSUP;
+	}
+
+	for (int i = 0; i < ARRAY_SIZE(command_line_prefix); i++) {
+		err = get_number(buf, &number);
+		if (err != 0) {
+			return -ENOTSUP;
+		}
+
+		if (!is_char(buf, ',')) {
+			return -ENOTSUP;
+		}
+
+		if (command_line_prefix[i] != number) {
+			return -ENOTSUP;
+		}
+	}
+
+	err = get_number(buf, &number);
+	if (err != 0) {
+		return -ENOTSUP;
+	}
+
+	if (!is_char(buf, '\r')) {
+		return -ENOTSUP;
+	}
+
+	if (number == 1) {
+		atomic_set_bit(ag->flags, BT_HFP_AG_CMER_ENABLE);
+		if ((ag->hf_features & BT_HFP_HF_FEATURE_3WAY_CALL) &&
+		    (ag->ag_features & BT_HFP_AG_FEATURE_3WAY_CALL)) {
+			LOG_DBG("Waiting for AT+CHLD=?");
+			return 0;
+		}
+
+		err = hfp_ag_next_step(ag, bt_hfp_ag_slc_connected, NULL);
+		return err;
+	} else if (number == 0) {
+		atomic_clear_bit(ag->flags, BT_HFP_AG_CMER_ENABLE);
+	} else {
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static void bt_hfp_ag_accept_cb(struct bt_hfp_ag *ag, void *user_data)
+{
+	struct bt_hfp_ag_call *call = (struct bt_hfp_ag_call *)user_data;
+
+	bt_hfp_ag_set_call_state(call, BT_HFP_CALL_ACTIVE);
+
+	if (bt_ag && bt_ag->accept) {
+		bt_ag->accept(call);
+	}
+}
+
+#if defined(CONFIG_BT_HFP_AG_3WAY_CALL)
+static int chld_release_all(struct bt_hfp_ag *ag)
+{
+	struct bt_hfp_ag_call *call;
+	bt_hfp_call_state_t call_state;
+
+	for (size_t index = 0; index < ARRAY_SIZE(ag->calls); index++) {
+		call = &ag->calls[index];
+
+		if (!atomic_test_bit(call->flags, BT_HFP_AG_CALL_IN_USING)) {
+			continue;
+		}
+
+		hfp_ag_lock(ag);
+		call_state = call->call_state;
+		hfp_ag_unlock(ag);
+
+		if ((call_state != BT_HFP_CALL_ACTIVE) ||
+		    ((call_state == BT_HFP_CALL_ACTIVE) &&
+		     atomic_test_bit(call->flags, BT_HFP_AG_CALL_INCOMING_HELD))) {
+			ag_terminate_call(call);
+		}
+	}
+
+	ag_notify_call_held_ind(ag, NULL);
+
+	return 0;
+}
+
+static void bt_hfp_ag_accept_other_cb(struct bt_hfp_ag *ag, void *user_data)
+{
+	struct bt_hfp_ag_call *call = (struct bt_hfp_ag_call *)user_data;
+	bt_hfp_call_state_t call_state;
+	int err;
+
+	if (!call) {
+		return;
+	}
+
+	hfp_ag_lock(ag);
+	call_state = call->call_state;
+	hfp_ag_unlock(ag);
```

Comment:
What if some other thread changes `call->call_state` right after you do `hfp_ag_unlock()` here? You'll still call `hfp_ag_update_indicator()`. Is that correct?

Reply:
And you'll also end up calling `bt_hfp_ag_set_call_state(call, BT_HFP_CALL_ACTIVE);` even if the call isn't active anymore (the fact that you're using locking and released the lock gives the impression that the value may change outside of this thread context)

Reply:
In theory, there is such an issue. Using hfp_ag_lock/hfp_ag_unlock cannot guarantee the consistency of the state in different contexts, unless all the code that modifies the state is executed in the same context.

Reply:
Updated